cdm/zarr/src/main/java/ucar/unidata/io/zarr/RandomAccessDirectory.java

```diff
@@ -0,0 +1,270 @@
+/*
+ * Copyright (c) 2021 University Corporation for Atmospheric Research/Unidata
+ * See LICENSE for license information.
+ */
+
+package ucar.unidata.io.zarr;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import thredds.inventory.*;
+import ucar.nc2.NetcdfFiles;
+import ucar.nc2.util.cache.FileCacheable;
+import ucar.unidata.io.RandomAccessFile;
+import ucar.unidata.io.spi.RandomAccessFileProvider;
+
+import java.io.*;
+import java.nio.channels.WritableByteChannel;
+import java.util.*;
+
+/**
+ * This class allows a directory structure to be read in memory as a single file.
+ * RandomAccessDirectory implemented a tree structure with files as leaves.
+ * It is read-only - writes should use the leaf RandomAccessFile write methods
+ */
+public class RandomAccessDirectory extends ucar.unidata.io.RandomAccessFile implements FileCacheable, Closeable {
+
+  private static final Logger logger = LoggerFactory.getLogger(RandomAccessDirectory.class);
+
+  protected List<RandomAccessDirectoryItem> children; // all files within the store
+
+  private RandomAccessFile currentFile; // file currently containing the file pointer
+
+  private long currentFileStartPos = -1; // start position of current file, relative to the directory/store
+
+  protected int bufferSize;
+
+  private static final String WRITES_NOT_IMPLEMENTED_MESSAGE =
+      "Method not implemented: writes are not implemented in RandomAccessDirectory";
+
+  public RandomAccessDirectory(String location) throws IOException {
+    this(location, RandomAccessFile.defaultBufferSize);
+  }
+
+  public RandomAccessDirectory(String location, int bufferSize) throws IOException {
+    super(bufferSize);
+    this.bufferSize = bufferSize;
+    this.location = location.replace("\\", "/"); // standardize path
+    this.readonly = true; // RandomAccessDirectory does not support writes
+
+    // build children list
```

Comment:
I ended up flattening the tree, but I'm not sure if it was the right move (it will probably become clear one way or the other as I work on the IOSP), and I might bring back the tree.... the tradeoff is basically that a lot of reads from the same file are faster with the flattened structure, but searching is faster with the tree. So it really depends on whether we'll be accessing different files or the same file more often.